pfcpiface/up4.go

```diff
@@ -767,53 +781,114 @@ func (up4 *UP4) allocateInternalApplicationID(app application) (uint8, error) {
-//nolint:unused
-func (up4 *UP4) releaseInternalApplicationID(appFilter applicationFilter) {
-	app := application{
-		appIP:     appFilter.srcIP,
-		appL4Port: appFilter.srcPortRange,
-		appProto:  appFilter.proto,
+func (up4 *UP4) unsafeReleaseInternalApplicationID(appFilter up4ApplicationFilter) {
+	allocated, exists := up4.applicationIDs[appFilter]
+	if exists {
+		up4.applicationIDsPool = append(up4.applicationIDsPool, allocated.id)
+		delete(up4.applicationIDs, appFilter)
 	}
+}
 
-	allocated, exists := up4.applicationIDs[app]
-	if exists {
-		delete(up4.applicationIDs, app)
-		up4.applicationIDsPool = append(up4.applicationIDsPool, allocated)
+func (up4 *UP4) addInternalApplicationIDAndGetP4rtEntry(pdr pdr) (*p4.TableEntry, uint8, error) {
+	up4.applicationMu.Lock()
+	defer up4.applicationMu.Unlock()
+
+	var appFilter up4ApplicationFilter
+	if pdr.IsUplink() {
+		appFilter = up4ApplicationFilter{
+			appIP:     pdr.appFilter.dstIP,
+			appL4Port: pdr.appFilter.dstPortRange,
+		}
+	} else if pdr.IsDownlink() {
+		appFilter = up4ApplicationFilter{
+			appIP:     pdr.appFilter.srcIP,
+			appL4Port: pdr.appFilter.srcPortRange,
+		}
+	}
+
+	appFilter.appProto = pdr.appFilter.proto
+
+	if up4Application, exists := up4.applicationIDs[appFilter]; exists {
+		// application already exists, increment ref count.
+		// since we use Set usedBy will not be incremented if
+		// application was already created for this UE session + PDR ID.
+		up4Application.usedBy.Add(internalAppReference{
+			pdr.fseID, pdr.pdrID,
+		})
+
+		return nil, up4Application.id, nil
+	}
+
+	newAppID, err := up4.unsafeAllocateInternalApplicationID()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	releaseID := func() {
+		up4.unsafeReleaseInternalApplicationID(appFilter)
+	}
+
+	up4Application := internalApp{
+		id: newAppID,
+		usedBy: set.NewSet(internalAppReference{
+			pdr.fseID, pdr.pdrID,
+		}),
+	}
+
+	applicationsEntry, err := up4.p4RtTranslator.BuildApplicationsTableEntry(pdr, up4.conf.SliceID, newAppID)
+	if err != nil {
+		releaseID()
+		return nil, 0, ErrOperationFailedWithReason("build P4rt table entry for Applications table", err.Error())
 	}
+
+	up4.applicationIDs[appFilter] = up4Application
+
+	return applicationsEntry, up4Application.id, nil
 }
 
-func (up4 *UP4) getOrAllocateInternalApplicationID(pdr pdr) (uint8, error) {
-	var app application
+func (up4 *UP4) removeInternalApplicationIDAndGetP4rtEntry(pdr pdr) (*p4.TableEntry, uint8) {
+	up4.applicationMu.Lock()
+	defer up4.applicationMu.Unlock()
+
+	var appFilter up4ApplicationFilter
 	if pdr.IsUplink() {
-		app = application{
+		appFilter = up4ApplicationFilter{
 			appIP:     pdr.appFilter.dstIP,
 			appL4Port: pdr.appFilter.dstPortRange,
 		}
 	} else if pdr.IsDownlink() {
-		app = application{
+		appFilter = up4ApplicationFilter{
 			appIP:     pdr.appFilter.srcIP,
 			appL4Port: pdr.appFilter.srcPortRange,
 		}
 	}
 
-	app.appProto = pdr.appFilter.proto
+	appFilter.appProto = pdr.appFilter.proto
```

Comment:
I already saw this in `addInternalApplicationIDAndGetP4rtEntry`. To reduce duplication, I suggest extracting into a function.